File: mobile/concord/lib/api/models/server_member_summary.dart
import 'json_utils.dart';
import 'public_profile_response.dart';

class ServerMemberSummary {
  const ServerMemberSummary({
    required this.user,
    required this.joinedAt,
    required this.isMuted,
    required this.timedOutUntil,
  });

  factory ServerMemberSummary.fromJson(Map<String, dynamic> json) {
    return ServerMemberSummary(
      user: PublicProfileResponse.fromJson(json.field('User') as Map<String, dynamic>),
      joinedAt: parseDateTime(json.field('JoinedAt')),
      isMuted: json.field('IsMuted') as bool? ?? false,
      timedOutUntil: parseNullableDateTime(json.field('TimedOutUntil')),
    );
  }

  final PublicProfileResponse user;
  final DateTime joinedAt;

  final bool isMuted;

  final DateTime? timedOutUntil;

  bool get isTimedOut => timedOutUntil != null && timedOutUntil!.isAfter(DateTime.now());

  ServerMemberSummary copyWith({
    PublicProfileResponse? user,
    bool? isMuted,
    DateTime? timedOutUntil,
    bool clearTimedOutUntil = false,
  }) {
    return ServerMemberSummary(
      user: user ?? this.user,
      joinedAt: joinedAt,
      isMuted: isMuted ?? this.isMuted,
      timedOutUntil: clearTimedOutUntil ? null : (timedOutUntil ?? this.timedOutUntil),
    );
  }
}

File: backend/Concord/Concord.Application/Models/ServerMemberSummary.cs
using System.Text.Json.Serialization;

namespace Concord.Application.Models;

public class ServerMemberSummary
{
    [JsonPropertyName("User")]
    public PublicProfileResponse User { get; set; } = null!;

    [JsonPropertyName("JoinedAt")]
    public DateTime JoinedAt { get; set; }

    [JsonPropertyName("IsMuted")]
    public bool IsMuted { get; set; }

    [JsonPropertyName("TimedOutUntil")]
    public DateTime? TimedOutUntil { get; set; }
}
